// videojs-mobile-ui/dist/videojs-mobile-ui.cjs.js
/*! @name videojs-mobile-ui @version 1.2.3 @license MIT */
'use strict';

var videojs = require('video.js');
var window = require('global/window');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var videojs__default = /*#__PURE__*/_interopDefaultLegacy(videojs);
var window__default = /*#__PURE__*/_interopDefaultLegacy(window);

var version = "1.2.3";

/**
 * @file touchOverlay.js
 * Touch UI component
 */

/** @import Player from 'video.js/dist/types/player' */

const Component = videojs__default["default"].getComponent('Component');
const dom = videojs__default["default"].dom || videojs__default["default"];

/**
 * The `TouchOverlay` is an overlay to capture tap events.
 *
 * @extends Component
 */
class TouchOverlay extends Component {
  /**
  * Creates an instance of the this class.
  *
  * @param  {Player} player
  *         The `Player` that this class should be attached to.
  *
  * @param  {options} [options]
  *         The key/value store of player options.
  */
  constructor(player, options) {
    super(player, options);
    this.seekSeconds = options.seekSeconds;
    this.tapTimeout = options.tapTimeout;
    this.taps = 0;

    // Add play toggle overlay
    this.addChild('playToggle', {});

    // Clear overlay when playback starts or with control fade
    player.on(['playing', 'userinactive'], e => {
      this.removeClass('show-play-toggle');
    });

    // A 0 inactivity timeout won't work here
    if (this.player_.options_.inactivityTimeout === 0) {
      this.player_.options_.inactivityTimeout = 5000;
    }

    /**
     * Debounced tap handler.
     * Seeks number of (taps - 1) * configured seconds to skip.
     * One tap is a non-op
     *
     * @param {Event} event
     */
    this.handleTaps_ = videojs__default["default"].fn.debounce(event => {
      const increment = (this.taps - 1) * this.seekSeconds;
      this.taps = 0;
      if (increment < 1) {
        return;
      }
      const rect = this.el_.getBoundingClientRect();
      const x = event.changedTouches[0].clientX - rect.left;

      // Check if double tap is in left or right area
      if (x < rect.width * 0.4) {
        this.player_.currentTime(Math.max(0, this.player_.currentTime() - increment));
        this.addClass('reverse');
      } else if (x > rect.width - rect.width * 0.4) {
        this.player_.currentTime(Math.min(this.player_.duration(), this.player_.currentTime() + increment));
        this.removeClass('reverse');
      } else {
        return;
      }

      // Remove play toggle if showing
      this.removeClass('show-play-toggle');

      // Remove and readd class to trigger animation
      this.setAttribute('data-skip-text', `${increment} ${this.localize('seconds')}`);
      this.removeClass('skip');
      window__default["default"].requestAnimationFrame(() => {
        this.addClass('skip');
      });
    }, this.tapTimeout);
    this.enable();
  }

  /**
   * Builds the DOM element.
   *
   * @return {Element}
   *         The DOM element.
   */
  createEl() {
    const el = dom.createEl('div', {
      className: 'vjs-touch-overlay',
      // Touch overlay is not tabbable.
      tabIndex: -1
    });
    return el;
  }

  /**
  * Debounces to either handle a delayed single tap, or a double tap
   *
   * @param {Event} event
   *        The touch event
   *
   */
  handleTap(event) {
    // Don't handle taps on the play button
    if (event.target !== this.el_) {
      return;
    }
    if (event.cancelable) {
      event.preventDefault();
    }
    this.taps += 1;
    if (this.taps === 1) {
      this.removeClass('skip');
      this.toggleClass('show-play-toggle');
    }
    this.handleTaps_(event);
  }

  /**
   * Enables touch handler
   */
  enable() {
    this.firstTapCaptured = false;
    this.on('touchend', this.handleTap);
  }

  /**
   * Disables touch handler
   */
  disable() {
    this.off('touchend', this.handleTap);
  }
}
Component.registerComponent('TouchOverlay', TouchOverlay);

/** @import Player from 'video.js/dist/types/player' */
/** @import Plugin from 'video.js/dist/types/plugin' */
/** @import {MobileUiOptions} from './plugin' */

/**
 * Sets up swiping to enter and exit fullscreen.
 *
 * @this {Plugin}
 * @param {Player} player
 *  The player to initialise on.
 * @param {MobileUiOptions} pluginOptions
 *  The options used by the mobile ui plugin.
 */
const initSwipe = (player, pluginOptions) => {
  const {
    swipeToFullscreen,
    swipeFromFullscreen
  } = pluginOptions.fullscreen;
  if (swipeToFullscreen) {
    player.addClass('using-fs-swipe-up');
  }
  if (swipeFromFullscreen) {
    player.addClass('using-fs-swipe-down');
  }
  let touchStartY = 0;
  let couldBeSwiping = false;
  const swipeThreshold = 30;

  /**
   * Monitor the possible start of a swipe
   *
   * @param {TouchEvent} e Triggering touch event
   */
  const onStart = e => {
    const isFullscreen = player.isFullscreen();
    if (!isFullscreen && !swipeToFullscreen || isFullscreen && !swipeFromFullscreen) {
      couldBeSwiping = false;
      return;
    }
    touchStartY = e.changedTouches[0].clientY;
    couldBeSwiping = true;
    player.tech_.el().style.transition = '';
  };

  /**
   * Monitor the movement of a swipe
   *
   * @param {TouchEvent} e Triggering touch event
   */
  const onMove = e => {
    if (!couldBeSwiping) {
      return;
    }
    const currentY = e.touches[0].clientY;
    const deltaY = touchStartY - currentY;
    const isFullscreen = player.isFullscreen();
    let scale = 1;
    if (!isFullscreen && deltaY > 0) {
      // Swiping up to enter fullscreen: Zoom in (Max 1.1)
      scale = 1 + Math.min(0.1, deltaY / 500);
      player.tech_.el().style.transform = `scale(${scale})`;
    } else if (isFullscreen && deltaY < 0) {
      // Swiping down to exit fullscreen: Zoom out (Min 0.9)
      scale = 1 - Math.min(0.1, Math.abs(deltaY) / 500);
      player.tech_.el().style.transform = `scale(${scale})`;
    }
  };

  /**
   * Monitor the touch end to determine a valid swipe
   *
   * @param {TouchEvent} e Triggering touch event
   */
  const onEnd = e => {
    if (!couldBeSwiping) {
      return;
    }
    couldBeSwiping = false;
    player.tech_.el().style.transition = 'transform 0.3s ease-out';
    player.tech_.el().style.transform = 'scale(1)';
    if (e.type === 'touchcancel') {
      return;
    }
    const touchEndY = e.changedTouches[0].clientY;
    const deltaY = touchStartY - touchEndY;
    if (deltaY > swipeThreshold && !player.isFullscreen()) {
      player.requestFullscreen().catch(err => {
        player.log.warn('Browser refused fullscreen', err);
      });
    } else if (deltaY < -swipeThreshold && player.isFullscreen()) {
      player.exitFullscreen();
    }
  };
  player.el().addEventListener('touchstart', onStart, {
    passive: true
  });
  player.el().addEventListener('touchmove', onMove, {
    passive: true
  });
  player.el().addEventListener('touchend', onEnd, {
    passive: true
  });
  player.el().addEventListener('touchcancel', onEnd, {
    passive: true
  });
  player.on('dispose', () => {
    player.el().removeEventListener('touchstart', onStart, {
      passive: true
    });
    player.el().removeEventListener('touchmove', onMove, {
      passive: true
    });
    player.el().removeEventListener('touchend', onEnd, {
      passive: true
    });
    player.el().removeEventListener('touchcancel', onEnd, {
      passive: true
    });
    player.tech_.el().style.transform = '';
    player.tech_.el().style.transition = '';
  });
};

/**
 * @typedef {Object} MobileUiOptions
 * @property {Object} [fullscreen]
 *  Options for fullscreen behaviours.
 * @property {boolean} [fullscreen.enterOnRotate]
 *  If the device is rotated, enter fullscreen.
 *  Default `true`.
 * @property {boolean} [fullscreen.exitOnRotate]
 *  If the device is rotated, exit fullscreen, unless `lockOnRotate` is used.
 *  Default `true`.
 * @property {boolean} [fullscreen.lockOnRotate]
 *  When going fullscreen in response to rotation (`enterOnRotate`), also lock the orientation (not supported by iOS).
 *  Default `true`.
 * @property {boolean} [fullscreen.lockToLandscapeOnEnter]
 *  When fullscreen is entered by any means, lock the orientation (not supported by iOS).
 *  Default `false`.
 * @property {boolean} [fullscreen.swipeToFullscreen]
 *  Swipe up to enter fullscreen.
 *  Default `false`.
 * @property {boolean} [fullscreen.swipeFromFullscreen]
 *  Swipe down to exit fullscreen.
 *  Won't do anything on iOS native fullscreen, which has its own swipe down exit gesture.
 *  Default `false`.
 * @property {boolean} [fullscreen.disabled]
 *  All fullscreen functionality provided by this plugin disabled.
 *  Default `false`.
 * @property {Object} [touchControls]
 *  Options for tap overlay.
 * @property {number} [touchControls.seekSeconds]
 *  Increment to seek in seconds.
 *  Default `10`.
 * @property {number} [touchControls.tapTimeout]
 *  Timeout to consider multiple taps as double rather than two single.
 *  Default `300`.
 * @property {boolean} [touchControls.disableOnEnd]
 *  Disable the touch overlay when the video ends.
 *  Useful if an end screen overlay is used to avoid conflict.
 *  Default `false`.
 * @property {boolean} [touchControls.disabled]
 *  All tap overlay functionality provided by this plugin disabled.
 *  Default `false`.
 * @internal
 * @property {boolean} [forceForTesting]
 *  Used in unit tests
 */

/** @type {MobileUiOptions} */
const defaults = {
  fullscreen: {
    enterOnRotate: true,
    exitOnRotate: true,
    lockOnRotate: true,
    lockToLandscapeOnEnter: false,
    swipeToFullscreen: false,
    swipeFromFullscreen: false,
    disabled: false
  },
  touchControls: {
    seekSeconds: 10,
    tapTimeout: 300,
    disableOnEnd: false,
    disabled: false
  }
};
const screen = window__default["default"].screen;

/**
 * Gets 'portrait' or 'lanscape' from the two orientation APIs
 *
 * @return {string} orientation
 */
const getOrientation = () => {
  if (screen) {
    // Prefer the string over angle, as 0° can be landscape on some tablets
    const orientationString = ((screen.orientation || {}).type || screen.mozOrientation || screen.msOrientation || '').split('-')[0];
    if (orientationString === 'landscape' || orientationString === 'portrait') {
      return orientationString;
    }
  }

  // iOS only supports window.orientation
  if (typeof window__default["default"].orientation === 'number') {
    if (window__default["default"].orientation === 0 || window__default["default"].orientation === 180) {
      return 'portrait';
    }
    return 'landscape';
  }
  return 'portrait';
};

/**
 * Add UI and event listeners
 *
 * @function onPlayerReady
 * @param    {Player} player
 *           A Video.js player object.
 *
 * @param    {MobileUiOptions} [options={}]
 *           A plain object containing options for the plugin.
 */
const onPlayerReady = (player, options) => {
  player.addClass('vjs-mobile-ui');
  if (!options.touchControls.disabled) {
    if (options.touchControls.disableOnEnd || typeof player.endscreen === 'function') {
      player.addClass('vjs-mobile-ui-disable-end');
    }

    // Insert before the control bar
    const controlBarIdx = player.children_.indexOf(player.getChild('ControlBar'));
    player.touchOverlay = player.addChild('TouchOverlay', options.touchControls, controlBarIdx);
  }
  if (options.fullscreen.disabled) {
    return;
  }
  if (options.fullscreen.swipeToFullscreen || options.fullscreen.swipeFromFullscreen) {
    initSwipe(player, options);
  }
  let locked = false;
  const rotationHandler = () => {
    const currentOrientation = getOrientation();
    if (currentOrientation === 'landscape' && options.fullscreen.enterOnRotate) {
      if (!player.paused() && !player.isFullscreen()) {
        player.requestFullscreen().catch(err => {
          player.log.warn('Browser refused fullscreen request:', err);
        });
        if ((options.fullscreen.lockOnRotate || options.fullscreen.lockToLandscapeOnEnter) && screen.orientation && screen.orientation.lock) {
          screen.orientation.lock('landscape').then(() => {
            locked = true;
          }).catch(err => {
            videojs__default["default"].log.warn('Browser refused orientation lock:', err);
          });
        }
      }
    } else if (currentOrientation === 'portrait' && options.fullscreen.exitOnRotate && !locked) {
      if (player.isFullscreen()) {
        player.exitFullscreen();
      }
    }
  };
  if (options.fullscreen.enterOnRotate || options.fullscreen.exitOnRotate) {
    if (videojs__default["default"].browser.IS_IOS) {
      window__default["default"].addEventListener('orientationchange', rotationHandler);
      player.on('dispose', () => {
        window__default["default"].removeEventListener('orientationchange', rotationHandler);
      });
    } else if (screen.orientation) {
      // addEventListener('orientationchange') is not a user interaction on Android
      screen.orientation.onchange = rotationHandler;
      player.on('dispose', () => {
        screen.orientation.onchange = null;
      });
    }
  }
  player.on('fullscreenchange', _ => {
    player.log('fullscreenchange', player.isFullscreen(), options.fullscreen.lockToLandscapeOnEnter, getOrientation());
    if (player.isFullscreen() && options.fullscreen.lockToLandscapeOnEnter && getOrientation() === 'portrait') {
      screen.orientation.lock('landscape').then(() => {
        locked = true;
      }).catch(e => {
        videojs__default["default"].log('Browser refused orientation lock:', e);
      });
    } else if (!player.isFullscreen() && locked) {
      screen.orientation.unlock();
      locked = false;
    }
  });
  player.on('ended', _ => {
    if (locked === true) {
      screen.orientation.unlock();
      locked = false;
    }
  });
};

/**
 * Adds a mobile UI for player control, and fullscreen orientation control
 *
 * @function mobileUi
 * @param    {MobileUiOptions} [options={}] Plugin options
 */
const mobileUi = function (options = {}) {
  if (options.forceForTesting || videojs__default["default"].browser.IS_ANDROID || videojs__default["default"].browser.IS_IOS) {
    this.ready(() => {
      onPlayerReady(this, videojs__default["default"].obj.merge(defaults, options));
    });
  }
};

// Register the plugin with video.js.
videojs__default["default"].registerPlugin('mobileUi', mobileUi);

// Include the version number.
mobileUi.VERSION = version;

module.exports = mobileUi;
